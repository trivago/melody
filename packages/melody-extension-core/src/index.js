/**
 * Copyright 2017 trivago N.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS-IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { unaryOperators, binaryOperators, tests } from './operators';
import { AutoescapeParser } from './parser/autoescape';
import { BlockParser } from './parser/block';
import { DoParser } from './parser/do';
import { EmbedParser } from './parser/embed';
import { ExtendsParser } from './parser/extends';
import { FilterParser } from './parser/filter';
import { FlushParser } from './parser/flush';
import { ForParser } from './parser/for';
import { FromParser } from './parser/from';
import { IfParser } from './parser/if';
import { ImportParser } from './parser/import';
import { IncludeParser } from './parser/include';
import { MacroParser } from './parser/macro';
import { SetParser } from './parser/set';
import { SpacelessParser } from './parser/spaceless';
import { UseParser } from './parser/use';
import { MountParser } from './parser/mount';

import forVisitor from './visitors/for';
import testVisitor from './visitors/tests';
import filters from './visitors/filters';
import functions from './visitors/functions';

const filterMap = [
    'attrs',
    'classes',
    'styles',
    'batch',
    'escape',
    'format',
    'merge',
    'nl2br',
    'number_format',
    'raw',
    'replace',
    'reverse',
    'round',
    'striptags',
    'title',
    'url_encode',
    'trim',
].reduce((map, filterName) => {
    map[filterName] = 'melody-runtime';
    return map;
}, Object.create(null));

Object.assign(filterMap, filters);

const functionMap = [
    'attribute',
    'constant',
    'cycle',
    'date',
    'max',
    'min',
    'random',
    'range',
    'source',
    'template_from_string',
].reduce((map, functionName) => {
    map[functionName] = 'melody-runtime';
    return map;
}, Object.create(null));
Object.assign(functionMap, functions);

export const extension = {
    tags: [
        AutoescapeParser,
        BlockParser,
        DoParser,
        EmbedParser,
        ExtendsParser,
        FilterParser,
        FlushParser,
        ForParser,
        FromParser,
        IfParser,
        ImportParser,
        IncludeParser,
        MacroParser,
        SetParser,
        SpacelessParser,
        UseParser,
        MountParser,
    ],
    unaryOperators,
    binaryOperators,
    tests,
    visitors: [forVisitor, testVisitor],
    filterMap,
    functionMap,
};

export {
    AutoescapeBlock,
    BlockStatement,
    BlockCallExpression,
    MountStatement,
    DoStatement,
    EmbedStatement,
    ExtendsStatement,
    FilterBlockStatement,
    FlushStatement,
    ForStatement,
    ImportDeclaration,
    FromStatement,
    IfStatement,
    IncludeStatement,
    MacroDeclarationStatement,
    VariableDeclarationStatement,
    SetStatement,
    SpacelessBlock,
    AliasExpression,
    UseStatement,
    UnaryNotExpression,
    UnaryNeqExpression,
    UnaryPosExpression,
    BinaryOrExpression,
    BinaryAndExpression,
    BitwiseOrExpression,
    BitwiseXorExpression,
    BitwiseAndExpression,
    BinaryEqualsExpression,
    BinaryNotEqualsExpression,
    BinaryLessThanExpression,
    BinaryGreaterThanExpression,
    BinaryLessThanOrEqualExpression,
    BinaryGreaterThanOrEqualExpression,
    BinaryNotInExpression,
    BinaryInExpression,
    BinaryMatchesExpression,
    BinaryStartsWithExpression,
    BinaryEndsWithExpression,
    BinaryRangeExpression,
    BinaryAddExpression,
    BinaryMulExpression,
    BinaryDivExpression,
    BinaryFloorDivExpression,
    BinaryModExpression,
    BinaryPowerExpression,
    BinaryNullCoalesceExpression,
    TestEvenExpression,
    TestOddExpression,
    TestDefinedExpression,
    TestSameAsExpression,
    TestNullExpression,
    TestDivisibleByExpression,
    TestConstantExpression,
    TestEmptyExpression,
    TestIterableExpression,
} from './types';
