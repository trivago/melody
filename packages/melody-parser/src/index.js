/**
 * Copyright 2017 trivago N.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS-IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import Parser from './Parser';
import TokenStream from './TokenStream';
import * as Types from './TokenTypes';
import Lexer from './Lexer';
import { EOF, CharStream } from './CharStream';
import { LEFT, RIGHT } from './Associativity';
import {
    setStartFromToken,
    setEndFromToken,
    copyStart,
    copyEnd,
    copyLoc,
    getNodeSource,
    createNode,
    hasTagStartTokenTrimLeft,
    hasTagEndTokenTrimRight,
    isMelodyExtension,
} from './util';

function parse(code, options, ...extensions) {
    return createExtendedParser(code, options, ...extensions).parse();
}

function createExtendedParser(code, options, ...extensions) {
    let passedOptions = options;
    const passedExtensions = extensions;
    if (isMelodyExtension(options)) {
        // Variant without options parameter: createExtendedParser(code, ...extensions)
        passedOptions = undefined;
        passedExtensions.unshift(options);
    }
    const lexer = createExtendedLexer(code, options, ...passedExtensions);
    const parser = new Parser(
        new TokenStream(lexer, passedOptions),
        passedOptions
    );
    for (const ext of passedExtensions) {
        parser.applyExtension(ext);
    }
    return parser;
}

function createExtendedLexer(code, options, ...extensions) {
    let passedOptions = options;
    const passedExtensions = extensions;
    if (isMelodyExtension(options)) {
        // Variant without options parameter: createExtendedLexer(code, ...extensions)
        passedOptions = undefined;
        passedExtensions.unshift(options);
    }
    const lexer = new Lexer(new CharStream(code), passedOptions);
    for (const ext of passedExtensions) {
        lexer.applyExtension(ext);
    }
    return lexer;
}

export {
    Parser,
    TokenStream,
    Lexer,
    EOF,
    CharStream,
    LEFT,
    RIGHT,
    parse,
    createExtendedLexer,
    createExtendedParser,
    setStartFromToken,
    setEndFromToken,
    copyStart,
    copyEnd,
    copyLoc,
    getNodeSource,
    createNode,
    hasTagStartTokenTrimLeft,
    hasTagEndTokenTrimRight,
    Types,
};
