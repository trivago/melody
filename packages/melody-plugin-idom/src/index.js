/**
 * Copyright 2017 trivago N.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS-IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import * as t from 'babel-types';
import mountVisitor from './visitors/mount.js';
import idomVisitor from './visitors/idom.js';

export default {
    visitors: [idomVisitor, mountVisitor],
    filterMap: {
        raw(path) {
            if (path.parentPath.is('PrintStatement')) {
                path.replaceWithJS(
                    t.expressionStatement(
                        t.callExpression(
                            t.identifier(
                                path.state.addImportFrom('melody-idom', 'raw')
                            ),
                            [path.get('target').node]
                        )
                    )
                );
            } else {
                path.replaceWithJS(
                    t.callExpression(
                        t.identifier(
                            path.state.addImportFrom('melody-idom', 'rawString')
                        ),
                        [path.get('target').node]
                    )
                );
            }
        },
    },
};
