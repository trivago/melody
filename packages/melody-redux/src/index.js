/**
 * Copyright 2017 trivago N.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS-IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import type { ReduxStore, Action } from 'melody-component';
import type { Component } from 'melody-component/component';
import { getParent } from 'melody-idom';
import { createWrapperComponent } from './WrapperComponent';

export type StateToPropsMapper = (state: Object, ownProps: Object) => Object;
export type DispatchToPropsMapper = (
    dispatch: (action: Action) => void
) => Object;
export type PropsMerger = (
    stateProps: Object,
    dispatchProps: Object,
    ownProps: Object
) => Object;

const findNextStore = comp => {
    for (let node = comp; node; node = getParent(node)) {
        if (node.store) {
            return node.store;
        }
    }
    if (process.env.NODE_ENV !== 'production') {
        console.error('Could not find store. Did you forget to provide a store?');
    }
    return null;
};

const defaultEmptyStateToPropsMapping = {};
const defaultMapStateToProps: StateToPropsMapper = () =>
    defaultEmptyStateToPropsMapping;
const defaultMapDispatchToProps: DispatchToPropsMapper = dispatch => ({
    dispatch,
});
const defaultMergeProps: PropsMerger = (
    stateProps,
    dispatchProps,
    parentProps
) => Object.assign({}, parentProps, stateProps, dispatchProps);

export function connect(
    mapStateToProps: ?StateToPropsMapper,
    mapDispatchToProps: ?(DispatchToPropsMapper | Object),
    mergeProps: ?PropsMerger
) {
    const shouldSubscribeToStore = !!mapStateToProps;
    const doStatePropsDependOnOwnProps =
        shouldSubscribeToStore && mapStateToProps.length !== 1;
    let finalMapDispatchToProps =
        mapDispatchToProps || defaultMapDispatchToProps;
    if (mapDispatchToProps && typeof mapDispatchToProps !== 'function') {
        const dispatchMap = mapDispatchToProps;
        finalMapDispatchToProps = dispatch =>
            Object.keys(dispatchMap).reduce((acc, key) => {
                acc[key] = (...args) => dispatch(dispatchMap[key](...args));
                return acc;
            }, {});
    }
    const doDispatchPropsDependOnOwnProps =
        finalMapDispatchToProps.length !== 1;
    const finalMapStateToProps = mapStateToProps || defaultMapStateToProps;
    const finalMergeProps = mergeProps || defaultMergeProps;

    return Component =>
        class ConnectWrapperComponent extends createWrapperComponent(
            Component
        ) {
            constructor() {
                super();
                this.store = null;
                this.storeConnection = null;

                this.renderProps = null;
                this.ownProps = null;
                this.stateProps = null;
                this.dispatchProps = null;

                this.mapStateToPropsFn = finalMapStateToProps;
                this.mapDispatchToPropsFn = finalMapDispatchToProps;

                if (process.env.NODE_ENV !== 'production') {
                    this.displayName = `connect`;
                }
            }

            apply(props: Object) {
                if (shouldSubscribeToStore && !this.storeConnection) {
                    this.subscribeToStore();
                }
                if (shallowEquals(this.ownProps, props)) {
                    return;
                }
                const store = this.store || findNextStore(this);
                if (this.ownProps !== props || this.renderProps === null) {
                    this.ownProps = props;
                    if (doStatePropsDependOnOwnProps || !this.stateProps) {
                        this.stateProps = this.mapStateToProps(
                            store.getState(),
                            props
                        );
                    }

                    if (
                        doDispatchPropsDependOnOwnProps ||
                        !this.dispatchProps
                    ) {
                        this.dispatchProps = this.mapDispatchToProps(
                            store.dispatch,
                            props
                        );
                    }

                    this.renderProps = finalMergeProps(
                        this.stateProps,
                        this.dispatchProps,
                        props
                    );
                }
                this.childInstance.apply(this.renderProps);
            }

            subscribeToStore() {
                if (shouldSubscribeToStore && !this.storeConnection) {
                    if (!this.store) {
                        this.store = findNextStore(this);
                    }
                    this.storeConnection = this.store.subscribe(() => {
                        const props = this.ownProps;
                        const newStateProps = this.mapStateToProps(
                            this.store.getState(),
                            props
                        );
                        if (newStateProps !== this.stateProps) {
                            const newRenderProps = finalMergeProps(
                                newStateProps,
                                this.dispatchProps,
                                props
                            );
                            this.stateProps = newStateProps;
                            const didRenderPropsChange = !shallowEquals(
                                this.renderProps,
                                newRenderProps
                            );
                            this.renderProps = newRenderProps;
                            if (didRenderPropsChange) {
                                this.childInstance.apply(this.renderProps);
                            }
                        }
                    });
                }
            }

            mapStateToProps(state, props) {
                let result = this.mapStateToPropsFn.call(null, state, props);
                if (typeof result === 'function') {
                    this.mapStateToPropsFn = result;
                    result = this.mapStateToPropsFn.call(null, state, props);
                }
                return result;
            }

            mapDispatchToProps(dispatch, props) {
                let result = this.mapDispatchToPropsFn.call(
                    null,
                    dispatch,
                    props
                );
                if (typeof result === 'function') {
                    this.mapDispatchToPropsFn = result;
                    result = this.mapDispatchToPropsFn.call(
                        null,
                        dispatch,
                        props
                    );
                }
                return result;
            }

            componentWillUnmount() {
                if (shouldSubscribeToStore) {
                    this.storeConnection();
                    this.storeConnection = null;
                    this.store = null;
                }
                this.mapStateToPropsFn = null;
                this.mapDispatchToPropsFn = null;
            }
        };
}

export function provide(store: ReduxStore, Component: Component): Component {
    return class StoreProviderComponent extends createWrapperComponent(
        Component
    ) {
        constructor() {
            super();
            this.store = store;

            if (process.env.NODE_ENV !== 'production') {
                this.displayName = `provide`;
            }
        }

        apply(props) {
            if (!shallowEquals(props, this.props)) {
                this.childInstance.apply(props);
            }
        }

        componentWillUnmount() {}
    };
}

const hasOwn = Object.prototype.hasOwnProperty;

// based on react-redux
function shallowEquals(a, b) {
    if (a === b) {
        return true;
    }

    if (!a || !b) {
        return false;
    }

    const keyOfA = Object.keys(a),
        keysOfB = Object.keys(b);

    if (keyOfA.length !== keysOfB.length) {
        return false;
    }

    for (let i = 0; i < keyOfA.length; i++) {
        if (!hasOwn.call(b, keyOfA[i]) || a[keyOfA[i]] !== b[keyOfA[i]]) {
            return false;
        }
    }

    return true;
}
