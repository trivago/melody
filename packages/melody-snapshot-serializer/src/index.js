import { Wrapper } from 'melody-test-utils';
import { getNodeData } from 'melody-idom';

const HTML_ELEMENT_REGEXP = /HTML\w*?Element/;
const isPlaceholderElement = val =>
    val !== undefined &&
    val !== null &&
    (val.nodeType === 1 || val.nodeType === 3 || val.nodeType === 8) &&
    val.constructor !== undefined &&
    val.constructor.name !== undefined &&
    HTML_ELEMENT_REGEXP.test(val.constructor.name) &&
    val.localName === 'm-placeholder' &&
    !!getNodeData(val);

const isWrapper = val =>
    val instanceof Wrapper || (!!val && val['isMelodyWrapper']);

export function print(val, serialize, indent) {
    if (isWrapper(val)) {
        if (val.elements.length > 1) {
            return serialize(val.elements);
        }
        return serialize(val.elements[0]);
    }
    const data = getNodeData(val);
    const inst = data.componentInstance;
    const propKeys = Object.keys(inst.props);
    const hasAttrs = propKeys.length > 0;
    if (!hasAttrs) {
        return `<${inst.displayName} />`;
    }
    const attrs = indent(
        propKeys
            .map(prop => `${prop}=${serialize(inst.props[prop])}`)
            .join('\n')
    );
    return `<${inst.displayName}\n${attrs}\n/>`;
}

export function test(val) {
    return isPlaceholderElement(val) || isWrapper(val);
}
