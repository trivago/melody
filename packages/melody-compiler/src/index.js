/**
 * Copyright 2017 trivago N.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS-IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { CharStream, Lexer, TokenStream, Parser } from 'melody-parser';
import { traverse, merge, Path, Scope } from 'melody-traverse';

import { Template, File } from './Template.js';
import analyse from './analyse/index.js';
import convert from './convert/index.js';
import State from './state/State.js';
import finalizer from './finalizer/index.js';

import * as t from 'babel-types';
import { CodeGenerator } from 'babel-generator';

// workaround for https://github.com/rollup/rollup/issues/430
import { TokenTypes } from 'melody-parser';
export { Template, File, TokenTypes };

export function compile(fileName: String, source: String, ...extensions) {
    const root = parseString(fileName, source, ...extensions);
    const template = new Template(root.expressions),
        file = new File(fileName, template),
        state = new State(file, source);
    let analyseVisitor = analyse,
        convertVisitor = convert;
    for (const ext of extensions) {
        if (ext.visitors) {
            if (Array.isArray(ext.visitors)) {
                for (const visitor of (ext.visitors: Array)) {
                    if (visitor.analyse) {
                        analyseVisitor = merge(analyseVisitor, visitor.analyse);
                    }
                    if (visitor.convert) {
                        convertVisitor = merge(convertVisitor, visitor.convert);
                    }
                }
            } else {
                if (ext.visitors.analyse) {
                    analyseVisitor = merge(
                        analyseVisitor,
                        ext.visitors.analyse
                    );
                }
                if (ext.visitors.convert) {
                    convertVisitor = merge(
                        convertVisitor,
                        ext.visitors.convert
                    );
                }
            }
        }
        if (ext.filterMap) {
            Object.assign(state.filterMap, ext.filterMap);
        }
        if (ext.functionMap) {
            Object.assign(state.functionMap, ext.functionMap);
        }
        if (ext.options) {
            Object.assign(state.options, ext.options);
        }
    }
    convertVisitor = merge(convertVisitor, finalizer);
    const scope = Scope.get(
        Path.get({
            container: file,
            key: 'template',
        })
    );
    traverse(file, analyseVisitor, scope, state);
    traverse(file, convertVisitor, scope, state);
    return t.file(file.template);
}

function parseString(fileName: string, source: string, ...extensions) {
    const lexer = new Lexer(new CharStream(source));
    for (const ext of extensions) {
        lexer.applyExtension(ext);
    }
    const parser = new Parser(new TokenStream(lexer));
    for (const ext of extensions) {
        parser.applyExtension(ext);
    }

    return parser.parse();
}

export function toString(jsAst, code) {
    const g = new CodeGenerator(jsAst, {}, code);
    // Babel sanitises strings to not contain complex characters
    // however we need them in order to be able to render complex strings
    g._generator.StringLiteral = function(node, parent) {
        var raw = this.getPossibleRaw(node);
        if (!this.format.minified && raw != null) {
            this.token(raw);
            return;
        }

        return this.token(JSON.stringify(node.value));
    };
    return g.generate();
}
