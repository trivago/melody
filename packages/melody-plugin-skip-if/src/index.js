/**
 * Copyright 2017 trivago N.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS-IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import * as t from 'babel-types';
import { Node, type, alias, visitor } from 'melody-types';
import template from 'babel-template';
import { Types, setStartFromToken, setEndFromToken } from 'melody-parser';

export class SkipIfBlock extends Node {
    constructor(target, expressions) {
        super();
        this.target = target;
        this.expressions = expressions;
    }
}
type(SkipIfBlock, 'SkipIfBlock');
alias(SkipIfBlock, 'Block');
visitor(SkipIfBlock, 'expressions');

export const SkipIfParser = {
    name: 'skip',
    parse(parser, token) {
        const tokens = parser.tokens;

        tokens.expect(Types.SYMBOL, 'if');
        const condition = tokens.expect(Types.SYMBOL);
        tokens.expect(Types.TAG_END);

        const skipIfBlock = new SkipIfBlock(
            condition,
            parser.parse((tokenText, token, tokens) => {
                return !!(
                    token.type === Types.TAG_START &&
                    tokens.nextIf(Types.SYMBOL, 'endskip')
                );
            }).expressions
        );

        setStartFromToken(skipIfBlock, token);
        setEndFromToken(skipIfBlock, tokens.expect(Types.TAG_END));

        return skipIfBlock;
    },
};

const buildSkipIfDefined = template(`
if (customElements.get(ELEMENT_NAME) !== undefined) {
    SKIP
} else {
    BODY
}
`);

export default {
    tags: [SkipIfParser],
    visitors: [
        {
            analyse: {
                SkipIfBlock(path) {
                    if (path.node.target.text === 'client') {
                        path.skip();
                    }
                },
            },
            convert: {
                SkipIfBlock: {
                    enter(path) {
                        if (path.node.target.text === 'client') {
                            path.replaceWithJS(skip(this));
                        }
                    },
                    exit(path) {
                        const target = path.node.target.text;
                        if (target === 'server') {
                            path.replaceWithMultipleJS(
                                ...path.node.expressions
                            );
                        } else if (target === 'defined') {
                            const el = path.findParentPathOfType('Element');
                            const elementName = el.node.name;
                            if (elementName.indexOf('-') === -1) {
                                this.error(
                                    'skip if defined can only be used inside of a custom element',
                                    el.node.elementNameLoc,
                                    `Custom Element must contain a "-" within their name but "${elementName}" does not seem to contain one.
More information about custom elements can be found here: https://developer.mozilla.org/en-US/docs/Web/Web_Components/Custom_Elements`,
                                    elementName.length
                                );
                            }
                            path.replaceWithJS(
                                buildSkipIfDefined({
                                    ELEMENT_NAME: t.stringLiteral(elementName),
                                    SKIP: skip(this),
                                    BODY: path.node.expressions,
                                })
                            );
                        } else {
                            this.error(
                                'Unknown skip if target',
                                path.node.target.pos,
                                `The target of the skip must be either 'client', 'server' or 'defined' but was "${target}" instead.`,
                                target.length
                            );
                        }
                    },
                },
            },
        },
    ],
};

function skip(state) {
    return t.expressionStatement(
        t.callExpression(
            t.identifier(state.addImportFrom('melody-idom', 'skip')),
            []
        )
    );
}
