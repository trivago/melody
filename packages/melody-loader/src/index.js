/**
 * Copyright 2017 trivago N.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS-IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { toString, compile } from 'melody-compiler';
import { extension as CoreExtension } from 'melody-extension-core';
import { getOptions } from 'loader-utils';
import { isString, isObject } from 'lodash';

module.exports = function loader(content) {
    this.cacheable();

    const loaderOptions = getOptions(this) || {
        plugins: [],
    };

    const args = [this.resourcePath, content, CoreExtension];
    if (loaderOptions.plugins) {
        for (const pluginName of loaderOptions.plugins) {
            if (isString(pluginName)) {
                try {
                    args.push(require('melody-plugin-' + pluginName));
                } catch (e) {
                    this.emitWarning(
                        'Could not find plugin ' +
                            pluginName +
                            '. Expected name to be melody-plugin-' +
                            pluginName
                    );
                }
            } else if (isObject(pluginName)) {
                args.push(pluginName);
            } else {
                this.emitWarning(
                    'Value passed as Melody plugin must be string or object. ' +
                        pluginName +
                        ' of type ' +
                        typeof pluginName +
                        ' was given'
                );
            }
        }
    }

    try {
        const result = toString(compile.apply(null, args), content);
        return result.code;
    } catch (e) {
        this.emitError(e);
        return (
            'import {text} from "melody-idom"; export default { render(options) { text("Could not load ' +
            this.resourcePath +
            '"); console.error("Could not load ' +
            this.resourcePath +
            '", ' +
            JSON.stringify(e.message) +
            '); } };'
        );
    }
};
