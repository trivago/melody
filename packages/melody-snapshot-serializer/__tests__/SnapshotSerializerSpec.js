import { getNodeData } from 'melody-idom';
import { Wrapper } from 'melody-test-utils';
import { test, print } from '../src';

describe('Snapshot serializer', () => {
    let prettyFormat;
    let indent;
    beforeEach(() => {
        prettyFormat = jest.fn(x => JSON.stringify(x));
        indent = jest.fn(x => '  ' + x);
    });

    it('should identify a Wrapper', () => {
        const el = document.createElement('el');
        const wrapper = new Wrapper(el);
        expect(test(wrapper)).toEqual(true);
    });

    it('should print without array when wrapping a single element', () => {
        const el = document.createElement('el');
        const wrapper = new Wrapper(el);
        print(wrapper, prettyFormat, indent);
        expect(prettyFormat).toHaveBeenCalledWith(wrapper.elements[0]);
    });

    it('should print an array when wrapping multiple elements', () => {
        const el = document.createElement('el');
        const el2 = document.createElement('el');
        const wrapper = new Wrapper([el, el2]);
        print(wrapper, prettyFormat, indent);
        expect(prettyFormat).toHaveBeenCalledWith(wrapper.elements);
    });

    it('should print shallow components', () => {
        const el = document.createElement('m-placeholder');
        getNodeData(el).componentInstance = {
            displayName: 'TestComponent',
            props: {},
        };
        expect(print(el, prettyFormat, indent)).toMatchSnapshot();
    });

    it('should print shallow components with props', () => {
        const el = document.createElement('m-placeholder');
        getNodeData(el).componentInstance = {
            displayName: 'TestComponent',
            props: {
                test: 1,
                foo: 'bar',
                baz: { hello: 'world' },
            },
        };
        expect(print(el, prettyFormat, indent)).toMatchSnapshot();
    });
});
