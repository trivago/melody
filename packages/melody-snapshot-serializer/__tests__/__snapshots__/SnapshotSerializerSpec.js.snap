// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot serializer should print shallow components 1`] = `"<TestComponent />"`;

exports[`Snapshot serializer should print shallow components with props 1`] = `
"<TestComponent
  test=1
foo=\\"bar\\"
baz={\\"hello\\":\\"world\\"}
/>"
`;
